sql/core/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SelectivePredicateHelper.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.PredicateHelper
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan}
+import org.apache.spark.sql.execution.FilterExec
+import org.apache.spark.sql.execution.columnar.InMemoryRelation
+
+/**
+ * [[InMemoryRelation]] is at sql/core module, so this helper trait also places in sql/core.
+ */
+trait SelectivePredicateHelper extends PredicateHelper {
+
+  /**
+   * Search a filtering predicate in a given logical plan
+   */
+  def hasSelectivePredicate(plan: LogicalPlan): Boolean = {
+    plan.exists {
+      // We do not actually need a selective predicate if the `InMemoryRelation` is materialized,
+      // because its statistics is correct enough. The call side would validate if the statistics
+      // is bigger than requirements.
+      case relation: InMemoryRelation if relation.isMaterialized => true
```

Comment:
So if the other side (filtering plan) has any `InMemoryRelation`, it will trigger pruning? If the cached plan is just (small) part of the filtering plan?

Reply:
It acutally only give a chance to optimize that the `InMemoryRelation` has no selective predicate but is a small relation.

According to the call side:
1. `PartitionPruning`, it use the overhead summary of leaf plan
     So if one of the leaf plan is a materialized `InMemoryRelation` with no selective predicate, the overhead summary will smaller than before and it likely can be optimized which can't before.

2. `InjectRuntimeFilter`, it use the maximum overhead of leaf plan
     So if one of the cached plan is small enough, it won't affect the final result. But if there is only one small cached plan, it likely can be optimized as well.